src/metabase/query_processor/parameters.clj

```diff
@@ -0,0 +1,252 @@
+      (let [{:keys [start end]} (absolute-date->range param-value)]
+        ["BETWEEN" field start end])
+      ;; relative date range
+      (condp = param-value
+        "past7days"  ["TIME_INTERVAL" field -7 "day"]
+        "past30days" ["TIME_INTERVAL" field -30 "day"]
+        "thisweek"   ["TIME_INTERVAL" field "current" "week"]
+        "thismonth"  ["TIME_INTERVAL" field "current" "month"]
+        "thisyear"   ["TIME_INTERVAL" field "current" "year"]
+        "lastweek"   ["TIME_INTERVAL" field "last" "week"]
+        "lastmonth"  ["TIME_INTERVAL" field "last" "month"]
+        "lastyear"   ["TIME_INTERVAL" field "last" "year"]
+        "yesterday"  ["=" field ["relative_datetime" -1 "day"]]
+        "today"      ["=" field ["relative_datetime" "current"]]))))
+
+(defn- merge-filter-clauses [base addtl]
+  (cond
+    (and (seq base)
+         (seq addtl)) ["AND" base addtl]
+    (seq base)        base
+    (seq addtl)       addtl
+    :else             []))
+
+(defn- expand-params-mbql [query-dict [{:keys [target value], :as param} & rest]]
+  (if param
+    (if (and param target value)
+      (let [filter-subclause (build-filter-clause param)
+            query            (assoc-in query-dict [:query :filter] (merge-filter-clauses (get-in query-dict [:query :filter]) filter-subclause))]
+        (expand-params-mbql query rest))
+      (expand-params-mbql query-dict rest))
+    query-dict))
+
+
+;;; +-------------------------------------------------------------------------------------------------------+
+;;; |                                           SQL QUERIES                                             |
+;;; +-------------------------------------------------------------------------------------------------------+
+
+
+(defn- extract-dates [value report-timezone]
+  (if-not (contains? relative-dates value)
+    ;; absolute date range
+    (absolute-date->range value)
+    ;; relative date range
+    (relative-date->range value report-timezone)))
+
+(defn- expand-date-range-param [report-timezone {[target param-name] :target, param-type :type, param-value :value, :as param}]
+  (if-not (= param-type "date")
+    param
+    (let [{:keys [start end]} (extract-dates param-value report-timezone)]
+      [(assoc param :target [target (str param-name ":start")], :value start)
+       (assoc param :target [target (str param-name ":end")],   :value end)])))
+
+(defn- substitute-param [param-name value query]
+  ;; TODO: escaping and protection against SQL injection!
+  (s/replace query (re-pattern (str "\\{\\{" param-name "\\}\\}")) value))
+
+(defn- substitute-all-params [query-dict [{:keys [value], [_ param-name] :target, :as param} & rest]]
+  (if param
+    (if-not (and param param-name value (string? param-name))
+      (substitute-all-params query-dict rest)
+      (let [query (update-in query-dict [:native :query] (partial substitute-param param-name value))]
+        (substitute-all-params query rest)))
+    query-dict))
```

Comment:
this would probably be clearer written as a `cond`.

```clojure
(defn- substitute-all-params [query-dict [{:keys [value], [_ param-name] :target, :as param} & rest]]
  (cond
    (not param)               query-dict
    (and (string? param-name)
         value)               (let [query (update-in query-dict [:native :query] (partial substitute-param param-name value))]
                                (substitute-all-params query rest))
    :else                     (substitute-all-params query-dict rest)))
```

At any rate `(and param param-name value (string? param-name))` can be simplified to 

```clojure
(and value (string? param-name))
```

since you already checked `param` in the line above and `(string? param-name)` will fail on `nil`